server/src/main/java/com/saralein/server/request/RequestParser.java

```diff
@@ -1,69 +1,71 @@
 
 public class RequestParser {
-    public Request parse(String request) throws Exception {
+    public Request parse(String rawRequest) throws Exception {
+        List<String> request = Arrays.asList(rawRequest.split(CRLF));
+        String[] requestLine = parseRequestLine(request);
+
+        return buildRequest(requestLine, request);
+    }
+
+    private String[] parseRequestLine(List<String> request) throws Exception {
         try {
-            List<String> fullRequest = Arrays.asList(split(request, CRLF));
-            HashMap<String, String> parsedRequest = createRequestMap(fullRequest);
-            return new Request(parsedRequest);
-        } catch(ArrayIndexOutOfBoundsException e) {
+            String requestLine = request.get(0);
+            return requestLine.split(" ");
+        } catch (ArrayIndexOutOfBoundsException e) {
             throw new Exception("Bad request. Connection closed.");
         }
     }
 
-    private HashMap<String, String> createRequestMap(List<String> fullRequest) {
-        HashMap<String, String> parsedRequest = addRequestLine(fullRequest);
-
-        if (fullRequest.size() > 0) {
-            parsedRequest.putAll(addHeadersAndBody(fullRequest));
-        }
-
-        return parsedRequest;
+    private Request buildRequest(String[] requestLine, List<String> request) {
+        return new Request.Builder()
+                .method(getMethod(requestLine))
+                .uri(getUri(requestLine))
+                .body(parseBody(request))
+                .addHeaders(parseHeaders(request))
+                .build();
     }
 
-    private HashMap<String, String> addRequestLine(List<String> fullRequest) {
-        String[] requestLine = splitRequestLine(fullRequest);
-
-        return new HashMap<String, String>(){{
-            put("method", parseMethod(requestLine));
-            put("uri", parseUri(requestLine));
-            put("version", parseVersion(requestLine));
-        }};
+    private String getMethod(String[] requestLine) {
+        return requestLine[0];
     }
 
-    private HashMap<String, String> addHeadersAndBody(List<String> fullRequest) {
-        HashMap<String, String> headersAndBody = new HashMap<>();
+    private String getUri(String[] requestLine) {
+        return requestLine[1];
+    }
 
-        fullRequest.stream()
-                .filter(line -> line.contains(":"))
-                .map(line -> split(line, ":"))
-                .forEach(line -> headersAndBody.put(line[0].trim(), line[1].trim()));
+    private String parseBody(List<String> request) {
+        String lastLine = request.get(request.size() - 1);
 
-        return headersAndBody;
+        return isBody(lastLine) ? getBodyValue(lastLine) : "";
     }
 
-    private String[] split(String request, String splitter) {
-        return request.split(splitter);
+    private HashMap<String, String> parseHeaders(List<String> request) {
+        return (HashMap<String, String>) request.stream()
```

Comment:
I think this is coming about because you are depending on a concretion instead of an abstraction. (It looks like it is stemming from the dependency on `HashMap<String, String>` in `Request`). What if you depended on `Map<String, String>` instead?

Reply:
In general, I will always try to depend on the collection interface instead of the implementation of it. For example, it looks like most places you are depending on `List<>` instead of `ArrayList<>`, but you are using an `ArrayList<>` to satisfy the `List<>` dependency.

Reply:
I realized after I made these comments that you mentioned you didn't want to make the Request more generic. Can you elaborate on why?

Reply:
Really I think I wasn't thinking it through.  It contradicts another thought I had yesterday looking at some other code where I was like, "Other people use the abstractions more than I do. I should work on that."  

I've updated it.  I will try to work on changing similar issues elsewhere when I see them.